pkg/egressgateway/manager.go

```diff
@@ -543,141 +538,96 @@ func (manager *Manager) updatePoliciesMatchedEndpointIDs() {
 	}
 }
 
-func (manager *Manager) updatePoliciesBySourceIP() {
-	manager.policyConfigsBySourceIP = make(map[string][]*PolicyConfig)
-
-	for _, policy := range manager.policyConfigs {
-		for _, ep := range policy.matchedEndpoints {
-			for _, epIP := range ep.ips {
-				ip := epIP.String()
-				manager.policyConfigsBySourceIP[ip] = append(manager.policyConfigsBySourceIP[ip], policy)
-			}
-		}
+func (manager *Manager) regenerateGatewayConfigs() {
+	for _, policyConfig := range manager.policyConfigs {
+		policyConfig.regenerateGatewayConfig(manager)
 	}
 }
 
-// policyMatches returns true if there exists at least one policy matching the
-// given parameters.
-//
-// This method takes:
-//   - a source IP: this is an optimization that allows to iterate only through
-//     policies that reference an endpoint with the given source IP
-//   - a callback function f: this function is invoked for each policy and for
-//     each combination of the policy's endpoints and destination/excludedCIDRs.
-//
-// The callback f takes as arguments:
-// - the given endpoint
-// - the destination CIDR
-// - a boolean value indicating if the CIDR belongs to the excluded ones
-// - the gatewayConfig of the  policy
-//
-// This method returns true whenever the f callback matches one of the endpoint
-// and CIDR tuples (i.e. whenever one callback invocation returns true)
-func (manager *Manager) policyMatches(sourceIP netip.Addr, f func(netip.Addr, netip.Prefix, bool, *gatewayConfig) bool) bool {
-	for _, policy := range manager.policyConfigsBySourceIP[sourceIP.String()] {
-		for _, ep := range policy.matchedEndpoints {
-			for _, endpointIP := range ep.ips {
-				if endpointIP != sourceIP {
-					continue
-				}
-
-				isExcludedCIDR := false
-				for _, dstCIDR := range policy.dstCIDRs {
-					if f(endpointIP, dstCIDR, isExcludedCIDR, &policy.gatewayConfig) {
-						return true
-					}
+// reconcileEgressRules takes all of the current manager state and
+// safely reconciles with the bpf map.
+func (manager *Manager) reconcileEgressRules() {
+	// Generate our desired state.
+	desired := map[egressmap.EgressPolicyKey4]egressmap.EgressPolicyVal4{}
```

Comment:
Now we have 2 `map[egressmap.EgressPolicyKey4]egressmap.EgressPolicyVal4{}`, desired and actual. We don't have to worry about the memory overhead that may cause?

Reply:
I really doubt this would cause memory issues, but if it's a major concern I could potentially try to memory profile the manager.